### PR TITLE
LossScaleOptimizer and Callback Support for MultiOptimizers.

### PR DESCRIPTION
**Pull request by gaga1313**
## Description
`MultiOptimizer` initial implementation did not support dynamic scaling, tracking Exponential Moving Average (EMA) variables, support for standard Keras callbacks (`ReduceLROnPlateau`, `LearningRateScheduler`, `SwapEMAWeights`, and `TensorBoard`). This PR enables MultiOptimizer support for dynamic gradient scaling using existing `LossScaleOptimizer`. It also adds support for Callback Level Dispatching modifying the callback classes themselves to be aware of the composite optimizers.

Added 6 robust test cases in `multi_optimizer_test.py` to test each modification:

- `test_learning_rate_scheduler`:
  - Verifies LearningRateScheduler correctly updates each sub-optimizer with a custom non-linear schedule function.
- `test_reduce_lr_on_plateau`:
  - Verifies ReduceLROnPlateau triggers properly and reduces the learning rates of all sub-optimizers proportionally once patience is exceeded.
- `test_tensorboard_logging`:
  - Verifies TensorBoard correctly collects and logs each sub-optimizer's learning rate as learning_rate_{opt_name} (and handles duplicates gracefully).
- `test_swap_ema_weights`:
  - Verifies SwapEMAWeights successfully swaps model sub-variables with their corresponding EMA weights.
- `test_multi_optimizer_blocks_inner_loss_scale_optimizer`
  - Verfies Verifies MultiOptimizer raises ValueError if inner optimizer is LSO.
- `test_multi_optimizer_wrapped_in_loss_scale_optimizer`
  - Verifies wrapping MultiOptimizer inside LossScaleOptimizer works as expected. 

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/callbacks/learning_rate_scheduler.py [hunk 1]
```diff
@@ -67,15 +67,36 @@ def on_epoch_begin(self, epoch, logs=None):
...
     def on_epoch_end(self, epoch, logs=None):
         logs = logs or {}
-        logs["learning_rate"] = float(
-            backend.convert_to_numpy(self.model.optimizer.learning_rate)
-        )
+        if hasattr(self.model.optimizer, "optimizers"):
+            for idx, opt in enumerate(self.model.optimizer.optimizers):
+                logs[f"learning_rate_{opt.name}_{idx}"] = float(
+                    backend.convert_to_numpy(opt.learning_rate)
+                )
```
Comment (hertschuh) [inline keras/src/callbacks/learning_rate_scheduler.py]: Remove the `_{idx}` part:
- the name is guaranteed to be unique
- the name is user controlled, `{idx}` is not, so it makes it much harder to find what you're looking for in the logs. If people are not happy with the auto-assigned name, they'll change it to something unique. However, they can't do that with `{idx}`.

File: keras/src/callbacks/learning_rate_scheduler.py [hunk 2]
```diff
@@ -67,15 +67,36 @@ def on_epoch_begin(self, epoch, logs=None):
...
+
+    def on_epoch_begin(self, epoch, logs=None):
+        if hasattr(self.model.optimizer, "optimizers"):
+            for idx, opt in enumerate(self.model.optimizer.optimizers):
+                learning_rate = self._update_optimizer_lr(opt, epoch)
+                if self.verbose > 0:
+                    io_utils.print_msg(
+                        f"\nEpoch {epoch + 1}: LearningRateScheduler setting "
+                        f"{opt.name}_{idx} learning rate to {learning_rate}."
+                    )
```
Comment (hertschuh) [inline keras/src/callbacks/learning_rate_scheduler.py]: Remove `_{idx}`, the name is unique, but the main reason is in my comment line 96.

File: keras/src/callbacks/reduce_lr_on_plateau.py [hunk 3]
```diff
@@ -108,21 +129,22 @@ def on_epoch_end(self, epoch, logs=None):
...
-                                f"learning rate to {new_lr}."
-                            )
+                    reduced = False
+                    if hasattr(self.model.optimizer, "optimizers"):
+                        for idx, opt in enumerate(
+                            self.model.optimizer.optimizers
+                        ):
+                            if self._reduce_optimizer_lr(
+                                opt, epoch, f"{opt.name}_{idx}"
+                            ):
```
Comment (hertschuh) [inline keras/src/callbacks/reduce_lr_on_plateau.py]: Remove `_{idx}`, the name is unique, but the main reason is in my comment line 96 in the other file.

File: keras/src/callbacks/tensorboard.py [hunk 4]
```diff
@@ -510,9 +510,19 @@ def _stop_trace(self, batch=None):
...
-        if isinstance(self.model.optimizer, Optimizer):
+        optimizer = self.model.optimizer
+        # Handles the case when the optimizer is wrapped by LossScaleOptimizer
+        if hasattr(optimizer, "inner_optimizer"):
+            optimizer = optimizer.inner_optimizer
+
+        if hasattr(optimizer, "optimizers"):
+            for idx, opt in enumerate(optimizer.optimizers):
+                logs[f"learning_rate_{opt.name}_{idx}"] = float(
+                    ops.convert_to_numpy(opt.learning_rate)
```
Comment (hertschuh) [inline keras/src/callbacks/tensorboard.py]: Remove `_{idx}`, the name is unique, but the main reason is in my comment line 96 in the other file.

File: keras/src/callbacks/reduce_lr_on_plateau.py [hunk 5]
```diff
@@ -85,9 +100,15 @@ def on_epoch_end(self, epoch, logs=None):
             # Delay setup until the model's metrics are all built
             self._set_monitor_op()
         logs = logs or {}
-        logs["learning_rate"] = float(
-            backend.convert_to_numpy(self.model.optimizer.learning_rate)
-        )
+        if hasattr(self.model.optimizer, "optimizers"):
+            for idx, opt in enumerate(self.model.optimizer.optimizers):
+                logs[f"learning_rate_{opt.name}_{idx}"] = float(
+                    backend.convert_to_numpy(opt.learning_rate)
```
Comment (hertschuh) [inline keras/src/callbacks/reduce_lr_on_plateau.py]: Remove `_{idx}`, the name is unique, but the main reason is in my comment line 96 in the other file.

File: keras/src/callbacks/reduce_lr_on_plateau.py [hunk 6]
```diff
@@ -77,6 +77,21 @@ def _reset(self):
...
+        if old_lr > np.float32(self.min_lr):
+            new_lr = old_lr * self.factor
+            new_lr = max(new_lr, self.min_lr)
+            optimizer.learning_rate = new_lr
+            if self.verbose > 0:
+                name_str = f" {name_suffix}" if name_suffix else ""
+                io_utils.print_msg(
+                    f"\nEpoch {epoch + 1}: ReduceLROnPlateau reducing"
+                    f"{name_str} learning rate to {new_lr}."
+                )
```
Comment (hertschuh) [inline keras/src/callbacks/reduce_lr_on_plateau.py]: nitpick: call the argument `name` instead of `name_suffix` because it's not really a suffix, you're just fixing the spacing.

File: keras/src/optimizers/multi_optimizer_test.py [hunk 7]
```diff
@@ -327,3 +328,337 @@ def test_multi_optimizer_with_model_fit(self):
...
+            w2.assign(new_trainable[1])
+        else:
+            lso_opt.apply(grads, [w1, w2])
+
+        self.assertAllClose(w1.numpy(), [[1.9, 1.9]])
+        self.assertAllClose(w2.numpy(), [[1.9, 1.9]])
+
+    @pytest.mark.requires_trainable_backend
+    def test_multi_optimizer_blocks_inner_loss_scale_optimizer(self):
+        """Verifies MultiOptimizer raises ValueError if
```
Comment (hertschuh) [inline keras/src/optimizers/multi_optimizer_test.py:638]: Rename the test and also verify MultiOptimizer within MultiOptimizer.

**Review by hertschuh — CHANGES_REQUESTED** [review]

File: keras/src/optimizers/multi_optimizer_test.py (hunk 7 above)
Comment (gaga1313) [reply keras/src/optimizers/multi_optimizer_test.py:638]: resolved 👍

**Review by gaga1313 — COMMENTED** [review]
resolved comments

**Review by hertschuh — APPROVED** [review]
Thanks!
